pkg/receive/hashring.go

```diff
@@ -0,0 +1,68 @@
+package receive
+
+import (
+	"errors"
+	"sort"
+
+	"github.com/improbable-eng/thanos/pkg/store/prompb"
+
+	"github.com/cespare/xxhash"
+)
+
+const sep = '\xff'
+
+// Hashring finds the correct host to handle a given time series
+// for a specified tenant.
+// It returns the hostname and any error encountered.
+type Hashring interface {
+	GetHost(string, *prompb.TimeSeries) (string, error)
+}
+
+// hash returns a hash for the given tenant and time series.
+func hash(tenant string, ts *prompb.TimeSeries) uint64 {
+	// Sort labelset to ensure a stable hash.
+	sort.Slice(ts.Labels, func(i, j int) bool { return ts.Labels[i].Name < ts.Labels[j].Name })
+
+	b := make([]byte, 0, 1024)
+	b = append(b, []byte(tenant)...)
+	for _, v := range ts.Labels {
+		b = append(b, v.Name...)
+		b = append(b, sep)
+		b = append(b, v.Value...)
+		b = append(b, sep)
+	}
+	return xxhash.Sum64(b)
+}
+
+// simpleHashring represents a group of hosts handling write requests.
+type simpleHashring []string
+
+// GetHost returns a hostname to handle the given tenant and time series.
+func (s simpleHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+	// Always return nil here to implement the Hashring interface.
+	return s[hash(tenant, ts)%uint64(len(s))], nil
+}
+
+// multiTenantHashring represents a set of Hashrings for different tenants.
+type multiTenantHashring map[string]Hashring
+
+// GetHost returns a hostname to handle the given tenant and time series.
+func (m multiTenantHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+	if h, ok := m[tenant]; ok {
+		return h.GetHost(tenant, ts)
+	}
+	if h, ok := m[""]; ok {
+		return h.GetHost(tenant, ts)
+	}
+	return "", errors.New("no matching hosts to handle tenant")
+}
+
+// NewHashring creates a multi-tenant hashring for a given
+// configuration that maps tenants to groups of hosts.
+func NewHashring(tenants map[string][]string) Hashring {
```

Comment:
let's architect this in a way right now where we instead of a direct map, have a list of filters that if all passed decides for that particular hashring, and for now just implement an exact-match-filter